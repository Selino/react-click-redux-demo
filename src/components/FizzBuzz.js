import React from "react"
import { useState, useEffect } from "react"
import { Form, Button, InputGroup, Container } from "react-bootstrap"

function processNumberArray(array) {
  const processedArray = array.map((item) => {
    if (item % 15 === 0) return "FizzBuzz"
    else if (item % 3 === 0) return "Fizz"
    else if (item % 5 === 0) return "Buzz"
    else return item
  })
  return processedArray
}

export default function FizzBuzz() {
  const [number, setNumber] = useState(2)
  const [numberArray, setNumberArray] = useState([])
  const [displayArray, setDisplayArray] = useState([""])

  useEffect(() => {
    console.log(numberArray)
    setDisplayArray(processNumberArray(numberArray))
  }, [numberArray])

  return (
    <Container>
      <Form
        onSubmit={(e) => {
          e.preventDefault()
          setNumberArray([...numberArray, number])
        }}
      >
        <InputGroup className='mb-3'>
          <Form.Control
            type='number'
            value={number}
            onChange={(e) => setNumber(e.target.value)}
          />
          <Button type='submit'>Submit</Button>
        </InputGroup>
      </Form>
      <div>{displayArray}</div>
    </Container>
  )
}
