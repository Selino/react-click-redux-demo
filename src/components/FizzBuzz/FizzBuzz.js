import React from "react"
import { useState, useEffect } from "react"
import {
  Form,
  Button,
  InputGroup,
  Container,
  Row,
  Col,
  Alert,
} from "react-bootstrap"
import { faCalculator } from "@fortawesome/free-solid-svg-icons"
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome"
import TransitionGroup from "react-transition-group/TransitionGroup"
import colors from "../../tokens/colors"
import fontWeights from "../../tokens/fontWeights"
import fontSizes from "../../tokens/fontSizes"

const numbersArray = [
  { num: 15, word: "FIZZBUZZ!", variant: "danger" },
  { num: 5, word: "Buzz", variant: "success" },
  { num: 3, word: "Fizz", variant: "warning" },
  { num: NaN, word: null, variant: "secondary" },
]

const getString = (array, number) => {
  const result = array.find(({ num }) => number % num === 0)
  return result === undefined ? null : result.word
}

const setVariant = (array, match) => {
  const result = array.find(({ word }) => word === match)
  return result.variant
}

export const AlertItem = (props) => {
  let x = {}
  switch (props.variant) {
    case "danger":
      x.color = colors.alertDanger
      x.backgroundColor = colors.alertDangerBackground
      break
    case "success":
      x.color = colors.alertSuccess
      x.backgroundColor = colors.alertSuccessBackground
      break
    case "warning":
      x.color = colors.alertWarning
      x.backgroundColor = colors.alertWarningBackground
      break
    default:
      x.color = colors.graysGray1
      x.backgroundColor = colors.graysGray4
      break
  }

  return (
    <Alert
      title={props.displayStatus}
      style={{
        color: x.color,
        backgroundColor: x.backgroundColor,
      }}
    >
      <Container>
        <Row>
          <Col
            className='display-status text-center text-md-left'
            style={{ fontWeight: fontWeights.medium, fontSize: fontSizes.h5 }}
            xs={12}
            sm={6}
          >
            {props.displayStatus}
          </Col>
          <Col
            className='text-center text-md-right result-number'
            style={{ fontWeight: fontWeights.bold, fontSize: fontSizes.h3 }}
          >
            {props.value >= 0 ? props.value : 0}
          </Col>
        </Row>
      </Container>
    </Alert>
  )
}

export const DisplayAlerts = (props) => {
  const alertsToDisplay = props.items
    .map((i, index) => {
      return (
        <AlertItem
          key={index}
          variant={setVariant(props.array, getString(props.array, i.num))}
          displayStatus={getString(props.array, i.num)}
          value={i.num}
        />
      )
    })
    .reverse()
  return <TransitionGroup>{alertsToDisplay}</TransitionGroup>
}

export default function FizzBuzz() {
  const [inputNumber, setInputNumber] = useState(1)
  const [displayArray, setDisplayArray] = useState([])
  const inputRef1 = React.createRef()

  const handleChange = (value) => {
    const regexp = new RegExp("^[0-9]+$")
    const cleanValue = value.match(regexp)

    cleanValue >= 0 &&
      cleanValue <= 99 &&
      // cleanValue != null &&
      setInputNumber(parseInt(cleanValue))
  }

  useEffect(() => {
    inputRef1.current.focus()
  }, [displayArray, inputRef1])

  return (
    <Container className='body-container'>
      <Row>
        <Col xs={12} sm={6} style={{ color: colors.black }}>
          <p>
            Enter a number between 1 & 99 to determine if it is a factor of
            (divides evenly into) 3 or 5. Get a "Fizz" if by 3, "Buzz" if by 5,
            and "<strong>FIZZBUZZ</strong>" if both!
          </p>

          <p>
            This is a classic JS algorithm that I've been asked to write more
            than once during white boarding interviews. I have not remembered it
            yet.{" "}
            <span role='img' aria-label='grimmace'>
              ðŸ˜¬
            </span>
          </p>
        </Col>
        <Col>
          <Form
            onSubmit={(e) => {
              e.preventDefault()
              setDisplayArray([...displayArray, { num: inputNumber }])
            }}
          >
            <InputGroup className='mb-3'>
              <Form.Control
                data-testid='number-input'
                type='number'
                value={inputNumber}
                onChange={(e) => {
                  handleChange(e.target.value)
                }}
                ref={inputRef1}
              />
              <InputGroup.Append>
                <Button
                  data-testid='submit-button'
                  type='submit'
                  style={{
                    color: colors.white,
                    backgroundColor: colors.primaryInactive,
                  }}
                >
                  <FontAwesomeIcon
                    alt='Calculate'
                    icon={faCalculator}
                  ></FontAwesomeIcon>
                </Button>
              </InputGroup.Append>
            </InputGroup>
          </Form>

          <div
            data-testid='list-area'
            style={{
              border: "solid 1px #CCC",
              borderRadius: "4px",
              height: "50vmin",
              overflow: "auto",
              padding: "1rem",
            }}
          >
            <DisplayAlerts array={numbersArray} items={displayArray} />
          </div>
        </Col>
      </Row>
    </Container>
  )
}
