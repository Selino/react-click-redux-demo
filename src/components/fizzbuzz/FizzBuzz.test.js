import React from "react"
import { shallow, mount, render } from "enzyme"
import FizzBuzz, { getString, AlertItem, DisplayAlerts } from "./FizzBuzz"

function range(start, end) {
  let foo = []
  for (let i = start; i <= end; i++) {
    foo.splice(i, 0, { num: i })
  }
  return foo
}

describe("getString - helper function", () => {
  it("return Fizz if % 3, Buzz if % 5, FIZZBUZZ! if both.", () => {
    const action = [15, 3, 5, 17].map((num) => getString(num))
    expect(action).toEqual(["FIZZBUZZ!", "Fizz", "Buzz", null])
  })
})

describe("AlertItem", () => {
  it("should render a single Alert item with displayStatus = Fizz, value = 3, and warning = variant.", () => {
    const wrapper = shallow(
      <AlertItem variant='warning' displayStatus='Fizz' value={3} />
    )
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find(".h2").text()).toEqual("3")
    expect(wrapper.find(".display-status").text()).toEqual("Fizz")
    expect(wrapper.prop("variant")).toEqual("warning")
  })

  it("should render a single Alert item with displayStatus = Buzz, value = 5, and warning = success.", () => {
    const wrapper = shallow(
      <AlertItem variant='success' displayStatus='Buzz' value={5} />
    )
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find(".h2").text()).toEqual("5")
    expect(wrapper.find(".display-status").text()).toEqual("Buzz")
    expect(wrapper.prop("variant")).toEqual("success")
  })

  it("should render a single Alert item with displayStatus = FIZZBUZZ!, value = 15, and warning = danger.", () => {
    const wrapper = shallow(
      <AlertItem variant='danger' displayStatus='FIZZBUZZ!' value={15} />
    )
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find(".h2").text()).toEqual("15")
    expect(wrapper.find(".display-status").text()).toEqual("FIZZBUZZ!")
    expect(wrapper.prop("variant")).toEqual("danger")
  })
})

describe("DisplayAlerts", () => {
  const testArray = range(1, 100)

  it("should render DisplayAlerts alert for full range of numbers within given range().", () => {
    const wrapper = shallow(<DisplayAlerts items={testArray} />)
    expect(wrapper.find(AlertItem).length).toBe(100)
  })

  it("should render DisplayAlerts alert for full range of numbers within given range().", () => {
    const wrapper = shallow(<DisplayAlerts items={testArray} />)
    expect(wrapper).toMatchSnapshot()
  })
})

describe("FizzBuzz", () => {
  // button push with number in it
  // set number < 1
  // set number > 99
  // push button with range(1-99) set
  let wrapper
  beforeEach(() => {
    wrapper = mount(<FizzBuzz />)
  })

  it("should render the full component with default value in field, no alerts", () => {
    const input = wrapper.find("input")
    const button = wrapper.find("Button")

    input.simulate("focus")
    input.simulate("change", { target: { value: "15" } })
    button.simulate("click")
    expect(wrapper).toMatchSnapshot()
  })
})
