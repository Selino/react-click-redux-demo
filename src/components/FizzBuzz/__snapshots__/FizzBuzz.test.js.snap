// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FizzBuzz should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="col-6"
      >
        <form
          class=""
        >
          <div
            class="mb-3 input-group"
          >
            <input
              class="form-control"
              data-testid="number-input"
              type="number"
              value="1"
            />
            <div
              class="input-group-append"
            >
              <button
                class="btn btn-primary"
                data-testid="submit-button"
                type="submit"
              >
                <svg
                  alt="Calculate"
                  aria-hidden="true"
                  class="svg-inline--fa fa-calculator fa-w-14 "
                  data-icon="calculator"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 448 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M400 0H48C22.4 0 0 22.4 0 48v416c0 25.6 22.4 48 48 48h352c25.6 0 48-22.4 48-48V48c0-25.6-22.4-48-48-48zM128 435.2c0 6.4-6.4 12.8-12.8 12.8H76.8c-6.4 0-12.8-6.4-12.8-12.8v-38.4c0-6.4 6.4-12.8 12.8-12.8h38.4c6.4 0 12.8 6.4 12.8 12.8v38.4zm0-128c0 6.4-6.4 12.8-12.8 12.8H76.8c-6.4 0-12.8-6.4-12.8-12.8v-38.4c0-6.4 6.4-12.8 12.8-12.8h38.4c6.4 0 12.8 6.4 12.8 12.8v38.4zm128 128c0 6.4-6.4 12.8-12.8 12.8h-38.4c-6.4 0-12.8-6.4-12.8-12.8v-38.4c0-6.4 6.4-12.8 12.8-12.8h38.4c6.4 0 12.8 6.4 12.8 12.8v38.4zm0-128c0 6.4-6.4 12.8-12.8 12.8h-38.4c-6.4 0-12.8-6.4-12.8-12.8v-38.4c0-6.4 6.4-12.8 12.8-12.8h38.4c6.4 0 12.8 6.4 12.8 12.8v38.4zm128 128c0 6.4-6.4 12.8-12.8 12.8h-38.4c-6.4 0-12.8-6.4-12.8-12.8V268.8c0-6.4 6.4-12.8 12.8-12.8h38.4c6.4 0 12.8 6.4 12.8 12.8v166.4zm0-256c0 6.4-6.4 12.8-12.8 12.8H76.8c-6.4 0-12.8-6.4-12.8-12.8V76.8C64 70.4 70.4 64 76.8 64h294.4c6.4 0 12.8 6.4 12.8 12.8v102.4z"
                    fill="currentColor"
                  />
                </svg>
              </button>
            </div>
          </div>
          <p>
            Enter a number between 1 & 99 to determine if it is a factor of (divides evenly into) 3 or 5. Get a "Fizz" if by 3, "Buzz" if by 5, and "
            <strong>
              FIZZBUZZ
            </strong>
            " if both!
          </p>
          <p>
            This is a classic JS algorithm that I've been asked to write more than once during white boarding interviews. I have not remembered it yet. 
            <span
              aria-label="grimmace"
              role="img"
            >
              ðŸ˜¬
            </span>
          </p>
        </form>
      </div>
      <div
        class="col"
      >
        <div
          data-testid="list-area"
          style="border: 1px solid #ccc; border-radius: 4px; overflow: auto; padding: 1rem;"
        >
          <div />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
